EMR Serverless Create/Stop/Delete Application deferrable mode

syedahsn opened the pull request:
This PR adds deferrable mode to `EmrServerlessCreateApplication`, `EmrServerlessStopApplication`, and `EmrServerlessDeleteApplication`. This required a bit of change in the hook code, but hopefully its clear why I did that. Tests are included

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

syedahsn committed:
Add deferrable option to EmrServerlessCreateApplicationOperator and EmrServerlessStopApplication Operator

syedahsn committed:
Add deferrable option to EmrServerlessCreateApplicationOperator and EmrServerlessStopApplication Operator

syedahsn committed:
Remove wait for jobs to be cancelled from hook

syedahsn committed:
Add deferrable option to EmrServerlessStopApplicationOperator and EmrServerlessDeleteApplication Operator

syedahsn committed:
Add deferrable option to EmrServerlessStopApplicationOperator and EmrServerlessDeleteApplication Operator

syedahsn committed:
Add unit tests for EmrServerlessApplicationTrigger

syedahsn committed:
Add deferrable mode to EmrServerlessDeleteApplicationOperator
Add unit tests

syedahsn committed:
Add deferrable mode to EmrServerlessDeleteApplicationOperator
Add unit tests

syedahsn committed:
Update documentation

potiuk committed:
Limit click until we fix mypy issues (#32413)

Click 8.1.4 breaks our mypy checks. The upper limit can be lifted when
the #32412 issue is resolved

syedahsn committed:
Conform to D205 style

syedahsn committed:
Fix spelling mistake

tirkarthi committed:
Updates health check endpoint to include dag_processor status. (#32382)

avinashpandeshwar committed:
Fix BigQuery transfer operators to respect project_id arguments (#32232)

pragnareddye commented on airflow/providers/amazon/aws/operators/emr.py:
```diff
@@ -1368,10 +1469,17 @@ class EmrServerlessDeleteApplicationOperator(EmrServerlessStopApplicationOperato
+        Default is 60 seconds.
+    :param deferrable: If True, the operator will wait asynchronously for application to be deleted.
+        This implies waiting for completion. This mode requires aiobotocore module to be installed.
+        (default: False)
```
Just keeping it consistent with other options
```suggestion
        Defaults to False
```

pragnareddye commented on airflow/providers/amazon/aws/operators/emr.py:
```diff
@@ -1368,10 +1469,17 @@ class EmrServerlessDeleteApplicationOperator(EmrServerlessStopApplicationOperato
+    :param waiter_check_interval_seconds: (deprecated) Number of seconds between polling the state
+        of the application. Defaults to 60 seconds.
+    :waiter_max_attempts: Number of times the waiter should poll the application to check the state.
+        Default is 25.
```
```suggestion
        Defaults to 25.
```

pragnareddye commented on airflow/providers/amazon/aws/operators/emr.py:
```diff
@@ -1368,10 +1469,17 @@ class EmrServerlessDeleteApplicationOperator(EmrServerlessStopApplicationOperato
+    :waiter_max_attempts: Number of times the waiter should poll the application to check the state.
+        Default is 25.
+    :param waiter_delay: Number of seconds between polling the state of the application.
+        Default is 60 seconds.
```
```suggestion
        Defaults to 60 seconds.
```

ephraimbuddy committed:
Refactor SetupTeardownContext Manager (#32235)

* Refactor SetupTeardownContext Manager

This refactor also includes preventing unreasonable usage of the context manager
and typing of the class

* Fix typing

* Use else

xuganyu96 committed:
Use WebEncoder to encode DagRun.conf in DagRun's list view (#32385)

* Use WebEncoder to encode DagRun.conf in DagRun's list view

* test case for json_f with WebEncoder

* Always use WebEncoder in web views

uranusjr commented on airflow/providers/amazon/aws/hooks/emr.py:
```diff
@@ -258,7 +258,8 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
         """
-        List all jobs in an intermediate state, cancel them, then wait for those jobs to reach terminal state.
+        List all jobs in an intermediate state, cancel them, then returns the number of
+        cancelled jobs.
```
```suggestion
        Cancel jobs in an intermediate state, and return the number of cancelled jobs.
```

We’re working on enforcing this docstring format soon.

eladkal committed:
Add explanation about tag creation and cleanup (#32406)

* Add explanation about tag creation and cleanup

Lee-W committed:
build(pre-commit): check deferrable default value (#32370)


---------

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

phanikumv committed:
Refactor Sqlalchemy queries to 2.0 style (Part 4) (#32339)

potiuk committed:
Cover wider set of excluded tests for re2 check (#32424)

We are adding more tests following *_test folder convention (for
example helm_tests) and re2 check should exclude all of those.

amoghrajesh committed:
Migrating helm chart tests to helm_tests to prevent test DB init (#32394)



---------

Co-authored-by: Amogh Desai <adesai@adesai-MBP16.local>

AlexisBRENON committed:
Document cron and delta timetables (#32392)

Add a comparison between delta and cron data interval timetable to
highlight differences when catchup is False.

amoghrajesh committed:
Migrating Google AutoML example_dags to sys tests (#32368)



---------

Co-authored-by: Amogh Desai <adesai@adesai-MBP16.local>

syedahsn committed:
Fix failing unit test
PR comment fixes

vandonr-amz committed:
Introduce a base class for aws triggers (#32274)

vandonr-amz committed:
Remove ability to specify arbitrary hook params in AWS RDS trigger (#32386)

vandonr-amz commented on airflow/providers/amazon/aws/hooks/emr.py:
```diff
@@ -284,13 +284,7 @@ def cancel_running_jobs(self, application_id: str, waiter_config: dict = {}):
-                states=list(self.JOB_INTERMEDIATE_STATES.union({"CANCELLING"})),
-                WaiterConfig=waiter_config,
-            )
+        return count
```
isn't that a breaking change ?
Also, if you were to go with it, you could remove the waiter_config param.

vandonr-amz commented on airflow/providers/amazon/aws/operators/emr.py:
```diff
@@ -1025,11 +1031,12 @@ def __init__(
         self.release_label = release_label
         self.job_type = job_type
-        self.wait_for_completion = wait_for_completion
+        self.wait_for_completion = False if deferrable else wait_for_completion
```
this is not really needed, is it ?
I was actually wondering with the new default if it'd make sense to defer only if `deferrable && wait_for_completion` ?

vandonr-amz commented on airflow/providers/amazon/aws/operators/emr.py:
```diff
@@ -1079,6 +1098,29 @@ def execute(self, context: Context) -> str | None:
         return application_id
 
+    def start_application_deferred(self, context, event=None):
+        if event["status"] == "success":
```
you could flip this if to reduce indentation

syedahsn committed:
Use AwsBaseWaiterTrigger for Triggers
Add wait_for_completion on cancel_running_jobs so it is backwards compatible
